docs/rfc/rfc-008-don't-panic.md

```diff
@@ -0,0 +1,133 @@
+the history of Tendermint, and as we embark on the project of
+stabilizing the package, we find ourselves in the right moment to
+reexamine our use of panics, and largely where panics happen in the
+code base.
+
+There are still some situations where panics are acceptable and
+desireable, but it's important that Tendermint, as a project, comes to
+consensus--perhaps in the text of this document--on the situations
+where it is acceptable to panic.
+
+### References
+
+- [Defer Panic and Recover](https://go.dev/blog/defer-panic-and-recover)
+- [Why Go gets exceptions right](https://dave.cheney.net/tag/panic)
+- [Don't panic](https://dave.cheney.net/practical-go/presentations/gophercon-singapore-2019.html#_dont_panic)
+
+## Discussion
+
+### Acceptable Panics
+
+#### Initialization
+
+It is unambiguously safe (and desireable) to panic in `init()`
+functions in response to any kind of error. These errors are caught by
+tests, and occur early enough in process initialization that they
+won't cause unexpected runtime crashes.
+
+Other code that is called early in process initialization MAY panic,
+in some situations if it's not possible to return an error or cause
+the process to abort early, although these situations should be
+vanishingly slim.
+
+#### Data Corruption
+
+If Tendermint code encounters an inconsistency that could be
+attributed to data corruption or a logical impossibility it is safer
+to panic and crash the process than continue to attempt to make
+progress in these situations.
+
+Examples including reading data out of the storage engine that
+is invalid or corrupt, or encountering an ambiguous situation where
+the process should halt. Generally these forms of corruption are
+detected after interacting with a trusted but external data source,
+and reflect situations where the author thinks its safer to terminate
+the process immediately rather than allow execution to continue.
+
+TODO: Add more and better examples here.
+
+#### Static Validity
+
+It is acceptable to panic for invariant violations, within a library
+or package, in situations that should be statically impossible,
+because there is no way to make these kinds of assertions at compile
+time.
+
+For example, type-asserting `interface{}` values returned by
+`container/list` and `container/heap` (and similar), is acceptable,
+because package authors should have exclusive control of the inputs to
+these containers. Packages should not expose the ability to add
+arbitrary values to these data structures.
+
+#### Controlled Panics Within Libraries
+
+In some algorithms with highly recursive structures or very nested
+call patterns, using a panic, in combination with conditional recovery
+handlers results in more manageable code. Ultimately this is a limited
+application, and implementations that use panics internally should
+only recover conditionally, filtering out panics rather than ignoring
+or handling all panics.
```

Comment:
I would be interested to see some more examples of this. To my mind, a panic is generally for when something is critically and unrecoverably wrong. The code cannot do its job and may need intervention before starting up again. If the issue can be handled, why not just return an error?